### PR TITLE
PERF: perform reductions block-wise

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -345,6 +345,37 @@ def _verify_integrity(self):
...
+        if self.ndim == 1:
+            # we'll be returning a scalar
+            blk = self.blocks[0]
+            return func(blk.values, *args, **kwargs)
+
+        res = {}
+        for blk in self.blocks:
+            bres = func(blk.values, *args, **kwargs)
+            if np.ndim(bres) == 0 and blk.shape[0] != 1:
+                # i.e. we reduced over all axes and not just one; re-do column-wise
```
Comment (TomAugspurger) [inline pandas/core/internals/managers.py]: I don't quite understand this case. How do we get here? re-calling `func` doesn't seem ideal.

Comment (jbrockmendel) [reply pandas/core/internals/managers.py]: IIRC it was when we have axis=None

Comment (jbrockmendel) [reply pandas/core/internals/managers.py]: after some digging, this appears to be coming from nanops funcs that are either not getting axis passed or are not handling it correctly.

Comment (jbrockmendel) [reply pandas/core/internals/managers.py]: updated to make this check unnecessary

**Comment by jbrockmendel** [comment]
Consolidated the conditions under which we go through the block-wise path.  In a follow-up, this will allow for significant cleanup of the non-blockwise path (i.e. existing _reduce code) because the set of cases it has to handle is cut down.

**Comment by jbrockmendel** [comment]
i hypothesize that after this and #30416 we'll be able to get rid of quite a bit of the existing special case handling in DataFrame._reduce.

**Comment by jbrockmendel** [comment]
docbuild failure looks unrelated

**Comment by jbrockmendel** [comment]
@jreback @TomAugspurger ideally id like to get this and #29941 in before the RC.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -7662,6 +7662,26 @@ def _get_data(axis_matters):
...
+            if axis == 1:
+                df = df.T
+                axis = 0
+
+            out_dtype = "bool" if filter_type == "bool" else None
+
+            # After possibly _get_data and transposing, we are now in the
+            #  simple case where we can use BlockManager._reduce
+            res = df._data.reduce(op, axis=1, skipna=skipna, **kwds)
+            assert isinstance(res, dict)
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Planning to keep these asserts in?

**Review by TomAugspurger — COMMENTED** [review]
Changes look nice. Any user-facing changes that need a whatsnew?

Do you think we have sufficient test coverage here?

**Comment by jbrockmendel** [comment]
> Any user-facing changes that need a whatsnew?

None that I've identified.

> Do you think we have sufficient test coverage here?

It seems pretty solid.

> Planning to keep these asserts in?

Fine by me either way.

**Comment by jreback** [comment]
ok by me (assuming you are going to back and de-gross / consolidate this code at some point).

**Comment by jbrockmendel** [comment]
> assuming you are going to back and de-gross / consolidate this code at some point

The code added here is is pretty de-grossed.  Some of the assertions could be removed to make it less verbose, but unless we get 2D EAs (which is a windmill ive mostly stopped tilting at), this is about as good as it gets.

That said, inside DataFrame._reduce below the code introduced here is a bunch of really gross code that we _can_ clean up after this.

**Comment by jreback** [comment]
k thanks

**Pull request merged**